clippy_lints/src/len_zero.rs

```diff
@@ -425,8 +425,12 @@ fn check_len(
     }
 }
 
-fn check_empty_expr(cx: &LateContext<'_>, span: Span, lit1: &Expr<'_>, lit2: &Expr<'_>, op: &str) {
+fn check_empty_expr(cx: &LateContext<'_>, span: Span, mut lit1: &Expr<'_>, lit2: &Expr<'_>, op: &str) {
     if (is_empty_array(lit2) || is_empty_string(lit2)) && has_is_empty(cx, lit1) {
+        if let ExprKind::Unary(UnOp::Deref, expr) = lit1.kind {
+            lit1 = expr;
+        }
```

Comment:
I think this only covers 1 deref, so 2 derefs or more wouldn't be covered, i.e.

```rust
    let s = "Hello, world!";
    let s1 = &&s;
    println!("{}", **s1 == "");
```

Reply:
I think getting the expression that is deref'd in a loop can be a solution

Reply:
That's one way to do it, but we have a util function that derefs expr

https://github.com/rust-lang/rust-clippy/blob/3cf2c17e53d115bd8c233a268c0d0ff0fbb2395e/clippy_utils/src/lib.rs#L2276-L2287

Maybe we can use that instead

Reply:
Another thing I noticed is that, parentheses aren't removed. So, something like the following:
```rs
**(s1) == ""
```
will become:
```rs
(s1).is_empty()
```
Is there any way to remove these redundant parentheses?

Reply:
We have another util to build suggestions, but it needs a little bit more digging:

https://github.com/rust-lang/rust-clippy/blob/78589ffad2c973f7863fdabc2df6e8ab2b54110f/clippy_utils/src/sugg.rs

IIRC, you can remove parentheses with this.

Reply:
Can you tell me what `default` parameter in `Sugg::hir_with_applicability` does?

Reply:
I found out that other lints don't remove the parentheses. Is this ok if I ignore parentheses?
AFAIK, the util you've suggested directly takes snippet from source code and has no method to remove unnecessary parentheses.

Reply:
I think it depends on the lint whether it will suggest removing the parentheses or not. Ideally, we'd have the parentheses removed.

But it's fine if we merge it without removing it in this PR, since the change is already an improvement as it is.

Let me know if you decide not to include the parentheses change in this PR.

Reply:
I think we can ignore parentheses now. Maybe this can be done in future?